### PR TITLE
Migrate CircleCI workflows to GitHub Actions (2/3)

### PR DESCRIPTION
**Pull request by AzfaarQureshi**
## Which problem is this solving?

As part of [issue](https://github.com/open-telemetry/opentelemetry-collector/issues/1234), this pull request migrates the loadtest, correctness and build-package jobs to GitHub Actions. `windows-msi` and the publish jobs will be migrated in PR 3/3

## Migration Plan

We suggest having CircleCI and GitHub Action jobs run in parallel for a few weeks. After the GitHub Actions jobs are running fine for a week or so and then remove the CircleCI workflows from config.yml
cc- @alolita, @shovnik

### CODE REVIEW

File: .github/workflows/build-and-test.yml [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+        with:
+          name: collector-binaries
+          path: ./bin
+  loadtestpre:
+    runs-on: ubuntu-latest
+    outputs:
+      matrix: ${{ steps.splitloadtest.outputs.matrix }}
+    steps:
+      - name: Checkout Repo
+        uses: actions/checkout@v2
```
Comment (AzfaarQureshi) [inline .github/workflows/build-and-test.yml:141]: `loadtestpre` does some pre-processing so we can simulate CircleCI testsharding on GitHub Actions. This cuts down the runtime of loadtests from 15mins to ~5mins
This job does the following: 

1. Get a list of all loadtest tests using `make testbed-list-loadtest`
2. Split tests into groups of two (`TestIdleMode|TestBallastMemory`, `TestLog10kDPS|TestMetric10kDPS` etc)
3. Create a JSON containing each test grouping. This JSON will be consumed by the [strategy.matrix](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategymatrix) of the loadtest job using [`fromJSON()`](https://github.blog/changelog/2020-04-15-github-actions-new-workflow-features/). The strategy.matrix is how we get to run the loadtests in parallel

File: .github/workflows/build-and-test.yml [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+              if [ -n "$curr" ] && (( i>1 )); then
+                MATRIX+=",{\"test\":\"$curr\"}"
+              elif [ -n "$curr" ]; then
+                MATRIX+="{\"test\":\"$curr\"}"
+              fi
+              curr="${TESTS[$i]}"
+            fi
+          done
+          MATRIX+="]}"
+          echo "::set-output name=matrix::$MATRIX"
```
Comment (AzfaarQureshi) [inline .github/workflows/build-and-test.yml:177]: This section of the code is what creates the strategy.matrix JSON

File: .github/workflows/build-and-test.yml [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+        run: make testbed-loadtest
+        env:
+          TEST_ARGS: "-test.run=${{ matrix.test }}"
+      - name: Create test result archive # some test results have invalid characters
+        if: ${{ failure() || success() }}
+        continue-on-error: true
+        run: tar -cvf test_results.tar testbed/tests/results
+      - name: Upload test results
+        if: ${{ failure() || success() }}
+        continue-on-error: true
```
Comment (AzfaarQureshi) [inline .github/workflows/build-and-test.yml:229]: - making a tarball of the test results because some tests have filenames with invalid characters (namely the `0*0bytes` test) which causes the upload action to fail. 

- `if: ${{ failure() || success() }}` this ensures that test results are uploaded even if the loadtest fails which can be useful for debugging

- `continue-on-error: true` Sometimes (rarely, but still) the upload action fails due to a network error on github's side. This makes sure the entire job doesnt fail if the test results fail to upload

File: testbed/testbed/receivers.go [hunk 4]
```diff
@@ -60,7 +60,7 @@ type DataReceiverBase struct {
 	Port int
 }
 
-const DefaultHost = "localhost"
+const DefaultHost = "127.0.0.1"
```
Comment (AzfaarQureshi) [inline testbed/testbed/receivers.go:63]: this line was causing an error with the fluentbit log test since `localhost` could not be resolved on the github actions runners.

File: testbed/tests/e2e_test.go [hunk 5]
```diff
@@ -42,7 +42,7 @@ func TestIdleMode(t *testing.T) {
 	)
 	defer tc.Stop()
 
-	tc.SetResourceLimits(testbed.ResourceSpec{ExpectedMaxCPU: 4, ExpectedMaxRAM: 50})
+	tc.SetResourceLimits(testbed.ResourceSpec{ExpectedMaxCPU: 4, ExpectedMaxRAM: 55})
```
Comment (AzfaarQureshi) [inline testbed/tests/e2e_test.go:45]: In the GitHub Action runners the RAM utilization is consistently higher (probably due to some other background processes on the GitHub machine). Loadtests don't pass unless these values are adjusted a little 😞

File: .github/workflows/build-and-test.yml [hunk 6]
```diff
@@ -45,6 +45,13 @@ jobs:
         uses: actions/setup-go@v2.1.3
         with:
           go-version: 1.15
+      - name: Restore module cache
+        uses: actions/cache@v2
+        env:
+          cache-name: cache-go-modules
+        with:
+          path: /home/runner/go/pkg/mod
+          key: go-pkg-mod-${{ runner.os }}-${{ hashFiles('./go.mod') }}
```
Comment (bogdandrutu) [inline .github/workflows/build-and-test.yml:54]: This can be split in a separate PR.

File: .github/workflows/build-and-test.yml [hunk 7]
```diff
@@ -125,3 +132,207 @@ jobs:
...
+        run: |
+          echo "GOPATH=$(go env GOPATH)" >> $GITHUB_ENV
+          echo "$(go env GOPATH)/bin" >> $GITHUB_PATH
+      - name: Restore module cache
+        uses: actions/cache@v2
+        env:
+          cache-name: cache-go-modules
+        with:
+          path: /home/runner/go/pkg/mod
+          key: go-pkg-mod-${{ runner.os }}-${{ hashFiles('./go.mod') }}
```
Comment (bogdandrutu) [inline .github/workflows/build-and-test.yml:210]: This code seem to be duplicated a lot between jobs, is there a way to avoid this and have a common step that all jobs will run?

Comment (AzfaarQureshi) [reply .github/workflows/build-and-test.yml:210]: Unfortunately I dont think there is a way to reuse steps across jobs as we can't use YAML anchors. It seems like the GitHub team is [aware](https://github.community/t/reusing-sharing-inheriting-steps-between-jobs-declarations/16851/14) this is a pain point though and working on an alternative feature soon :/

File: .github/workflows/build-and-test.yml (hunk 6 above)
Comment (AzfaarQureshi) [reply .github/workflows/build-and-test.yml:54]: 👍 sounds good I will remove it from this one

Comment (bogdandrutu) [reply .github/workflows/build-and-test.yml:54]: Please send an initial PR :)

Comment (AzfaarQureshi) [reply .github/workflows/build-and-test.yml]: Removed and PR is here: https://github.com/open-telemetry/opentelemetry-collector/pull/2305

File: .github/workflows/build-and-test.yml [hunk 8]
```diff
@@ -125,3 +125,207 @@ jobs:
...
+                MATRIX+="{\"test\":\"$curr\"}"
+              fi
+              curr="${TESTS[$i]}"
+            fi
+          done
+          MATRIX+="]}"
+          echo "::set-output name=matrix::$MATRIX"
+  loadtest:
+    runs-on: ubuntu-latest
+    needs: [setup-environment, loadtestpre]
```
Comment (bogdandrutu) [inline .github/workflows/build-and-test.yml:173]: This should need binaries as well correct?

File: .github/workflows/build-and-test.yml (hunk 2 above)
Comment (bogdandrutu) [reply .github/workflows/build-and-test.yml:170]: Can we have this as part of the setup? it is very simple work that I think we can execute in the setup to not restore the cache etc again.

**Review by bogdandrutu — APPROVED** [review]

File: .github/workflows/build-and-test.yml (hunk 8 above)
Comment (AzfaarQureshi) [reply .github/workflows/build-and-test.yml:173]: We thought that too but found out it doesn't actually use any of the binaries so we moved it out. It seems to be running test suites instead

File: .github/workflows/build-and-test.yml (hunk 2 above)
Comment (AzfaarQureshi) [reply .github/workflows/build-and-test.yml:170]: makes sense! moved!

File: .github/workflows/build-and-test.yml (hunk 8 above)
Comment (bogdandrutu) [reply .github/workflows/build-and-test.yml]: The test suite starts one of the binaries so it needs the binary

**Comment by bogdandrutu** [comment]
@AzfaarQureshi there is an error in the action
```
Error when evaluating 'strategy' for job 'loadtest'. (Line: 161, Col: 15): Error reading JToken from JsonReader. Path '', line 0, position 0.,(Line: 161, Col: 15): Unexpected value ''
```

File: .github/workflows/build-and-test.yml (hunk 8 above)
Comment (AzfaarQureshi) [reply .github/workflows/build-and-test.yml]: oh weird...I wonder why its passing when we're not including them. I will add it again. Thanks for the quick reviews!

Comment (bogdandrutu) [reply .github/workflows/build-and-test.yml]: It is re-building, in the makefile we depend on that step so it will just rebuild the binary :)

**Comment by bogdandrutu** [comment]
@AzfaarQureshi @shovnik I have couple more feedback about this, how can I give that to you? Is it ok for me to file issues and assign to you?

**Comment by AzfaarQureshi** [comment]
@bogdandrutu yup! that works 😄

**Comment by AzfaarQureshi** [comment]
@bogdandrutu the workflow actually ran faster (~[10 mins](https://github.com/open-telemetry/opentelemetry-collector/actions/runs/430892819)  when it didnt depend on the binaries from cross-compile as opposed to now when its running in [~20 mins](https://github.com/open-telemetry/opentelemetry-collector/actions/runs/431048926). Should we revert the change to what it was before?

**Comment by bogdandrutu** [comment]
@AzfaarQureshi yes please revert :D

File: .github/workflows/build-and-test.yml [hunk 9]
```diff
@@ -58,8 +60,29 @@ jobs:
...
+                MATRIX+="{\"test\":\"$curr\"}"
+              fi
+              curr="${TESTS[$i]}"
+            fi
+          done
+          MATRIX+="]}"
+          echo "::set-output name=matrix::$MATRIX"
       - name: Upload tool binaries
-        uses: actions/upload-artifact@v2
+        uses: actions/upload-artifact@v1
```
Comment (AzfaarQureshi) [inline .github/workflows/build-and-test.yml:85]: Turns out the flakiness of err CONNRESET when download/uploading artifacts is a known bug in `upload-artifact@v2` as noted [here](https://github.com/actions/upload-artifact/issues/71). reverting to `@v1` for now
